src/core/lib/iomgr/poller/eventmanager_libuv.cc

```diff
@@ -0,0 +1,77 @@
+/*
+ *
+ * Copyright 2019 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/iomgr/poller/eventmanager_libuv.h"
+
+#include <grpc/support/time.h>
+
+grpc::experimental::LibuvEventManager::Options::Options() : num_workers_(-1) {}
+grpc::experimental::LibuvEventManager::Options::Options(int num_workers)
+    : num_workers_(num_workers) {}
+
+grpc::experimental::LibuvEventManager::LibuvEventManager(const Options& options)
+    : options_(options), should_stop_(0), shutdown_refcount_(0) {
+  int num_workers = options_.num_workers();
+  // TODO(guantaol): replaces the hard-coded number with a flag.
+  if (num_workers < 0) num_workers = 32;
+
+  for (int i = 0; i < num_workers; i++) {
+    workers_.emplace_back(options_.thread_name_prefix().c_str(),
+                          &grpc::experimental::LibuvEventManager::RunWorkerLoop,
+                          this);
+    workers_.back().Start();
+  }
+}
+
+grpc::experimental::LibuvEventManager::~LibuvEventManager() {
+  Shutdown();
+  for (auto it = workers_.begin(); it != workers_.end(); it++) {
+    it->Join();
+  }
+}
+
+void grpc::experimental::LibuvEventManager::RunWorkerLoop(void* manager) {
+  LibuvEventManager* event_manager = static_cast<LibuvEventManager*>(manager);
+  while (true) {
+    if (event_manager->ShouldStop()) return;
+    gpr_sleep_until(gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
+                                 gpr_time_from_micros(10, GPR_TIMESPAN)));
+  }
+}
+
+bool grpc::experimental::LibuvEventManager::ShouldStop() {
+  return should_stop_.Load(grpc_core::MemoryOrder::ACQUIRE) != 0;
+}
+
+void grpc::experimental::LibuvEventManager::Shutdown() {
+  if (should_stop_.Load(grpc_core::MemoryOrder::ACQUIRE))
+    return;  // Already shut down.
+  while (shutdown_refcount_.Load(grpc_core::MemoryOrder::ACQUIRE) > 0)
```

Comment:
Let's do anything else besides an atomic loop. If this is run under fibers, this will never complete since this fiber will never yield and the other one will never get a chance to execute. Please convert this to a mu/cv combination.

Reply:
Sorry that my knowledge about fiber is minimal, but why will this never complete with fibers? Won't the two fibers run in parallel?

Reply:
Fibers use coöperative multithreading and they only yield when they perform a blocking operation (syscall, mutex, etc), so there's no guarantee that they run in parallel. Technically, even in standard C++ threads, there is no such guarantee of forward progress (somewhere deep in the heart of the spec). This problem is what forced us to disable the `channel_idle` filter, btw.

Reply:
Side note: even if it's not under fibers, keep in mind that cloud customers are paying per CPU unit that they use. A spin-loop costs them more $ than an idle CPU.